src/v/archival/segment_reupload.cc

```diff
@@ -193,25 +193,26 @@ segment_collector::find_next_compacted_segment(model::offset start_offset) {
     if (it == segment_set.end()) {
         vlog(
           archival_log.debug,
-          "Finding next compacted segment for {}: can't find segment after "
+          "Finding next segment for {}: can't find segment after "
           "offset: {}",
           _manifest.get_ntp(),
           start_offset);
         return {};
     }
 
     const auto& segment = *it;
-    if (segment->finished_self_compaction()) {
+    auto is_compacted = segment->finished_self_compaction();
+    if ((non_compacted && !is_compacted) || (!non_compacted && is_compacted)) {
```

Comment:
nit: maybe it doesn't fit the corner cases, but could this be simplified to `if (non_compacted != is_compacted)` or even something like`if (expected_compaction != segment->finished_self_compaction())`